native/python/zds_py.cpp

```diff
@@ -0,0 +1,210 @@
+        e2a_inplace(diag);
+        diag.pop_back();
+        throw std::runtime_error(diag);
+    }
+
+    for (auto &e : entries)
+    {
+        e2a_inplace(e.name);
+        e2a_inplace(e.dsorg);
+        e2a_inplace(e.volser);
+        e2a_inplace(e.recfm);
+    }
+    return entries;
+}
+
+std::string read_dataset(std::string dsn, std::string codepage)
+{
+    ZDS zds = {0};
+
+    if (!codepage.empty())
+    {
+        zds.encoding_opts.data_type = eDataTypeText;
+        strncpy(zds.encoding_opts.codepage, codepage.c_str(), sizeof(zds.encoding_opts.codepage) - 1);
+    }
+
+    a2e_inplace(dsn);
+    std::string response;
+    int rc = zds_read_from_dsn(&zds, dsn, response);
+
+    if (rc != 0)
+    {
+        std::string diag(zds.diag.e_msg, zds.diag.e_msg_len);
+        diag.push_back('\0');
+        e2a_inplace(diag);
+        diag.pop_back();
+        throw std::runtime_error(diag);
+    }
+
+    e2a_inplace(response);
+    return response;
+}
+
+std::string write_dataset(std::string dsn, std::string data, std::string codepage, std::string etag)
+{
+    ZDS zds = {0};
+
+    if (!codepage.empty())
+    {
+        zds.encoding_opts.data_type = eDataTypeText;
+        strncpy(zds.encoding_opts.codepage, codepage.c_str(), sizeof(zds.encoding_opts.codepage) - 1);
+    }
+
+    if (!etag.empty())
+    {
+        strncpy(zds.etag, etag.c_str(), sizeof(zds.etag) - 1);
+    }
+
+    a2e_inplace(dsn);
+    a2e_inplace(data);
+    int rc = zds_write_to_dsn(&zds, dsn, data);
+
+    if (rc != 0)
+    {
+        std::string diag(zds.diag.e_msg, zds.diag.e_msg_len);
+        diag.push_back('\0');
+        e2a_inplace(diag);
+        diag.pop_back();
+        throw std::runtime_error(diag);
+    }
+
+    std::string new_etag(zds.etag);
+    e2a_inplace(new_etag);
+    return new_etag;
+}
+
+void delete_dataset(std::string dsn)
+{
+    ZDS zds = {0};
+
+    a2e_inplace(dsn);
+    int rc = zds_delete_dsn(&zds, dsn);
+
+    if (rc != 0)
+    {
+        std::string diag(zds.diag.e_msg, zds.diag.e_msg_len);
+        diag.push_back('\0');
+        e2a_inplace(diag);
+        diag.pop_back();
+        throw std::runtime_error(diag);
+    }
+}
+
+/**
+ * Dataset Member Functions
+ */
+
+void create_member(std::string dsn)
+{
+    ZDS zds = {0};
+
+    a2e_inplace(dsn);
+
+    std::string empty_data = "";
+    a2e_inplace(empty_data);
+    int rc = zds_write_to_dsn(&zds, dsn, empty_data);
+
+    if (rc != 0)
+    {
+        std::string diag(zds.diag.e_msg, zds.diag.e_msg_len);
+        diag.push_back('\0');
+        e2a_inplace(diag);
+        diag.pop_back();
+        throw std::runtime_error(diag);
+    }
+}
+
+std::vector<ZDSMem> list_members(std::string dsn)
+{
+    std::vector<ZDSMem> members;
+    ZDS zds = {0};
+
+    a2e_inplace(dsn);
+    int rc = zds_list_members(&zds, dsn, members);
+
+    if (rc != 0 && rc != RTNCD_WARNING)
+    {
+        std::string diag(zds.diag.e_msg, zds.diag.e_msg_len);
+        diag.push_back('\0');
```

Comment:
Here's a shorter option for null-terminating the character array:

```suggestion
        std::string diag = std::string(zds.diag.e_msg) + "\0";
```

Although I am curious whether this is necessary if the array is filled with zeroes after the end of the message contents. You can verify this by checking whether `zds.diag.e_msg[zds.diag.e_msg_len]` is `0 (equivalent value for '\0')` 😋